src/core/createPackageLocator.js

```diff
@@ -0,0 +1,69 @@
+import path from 'path'
+import fs from 'fs'
+import { EOL } from 'os'
+
+class PackageLocator {
+  readPackageSync(location) {
+    return {
+      result: JSON.parse(fs.readFileSync(location, 'utf8')),
+      location,
+    }
+  }
+
+  readPackageUpSync(dirname) {
+    const locations = []
+
+    do {
+      try {
+        const location = path.join(dirname, 'package.json')
+        locations.push(location)
+
+        return this.readPackageSync(location)
+      } catch (err) {
+        // if not found, allopw search to continue
+        if (err.code !== 'ENOENT') throw err
+      }
+    } while (dirname !== (dirname = path.dirname(dirname)))
+
+    const notFoundError = new Error(`No such file or directory: ${EOL}${locations.join(EOL)}`)
+    notFoundError.code = 'ENOENT'
+
+    throw notFoundError
+  }
+}
+
+class CachedPackageLocator extends PackageLocator {
+  constructor() {
+    super()
+
+    this.store = {}
+    this.requests = new Map()
+  }
+
+  readPackageSync(location) {
+    const cached = this.store[location]
+    if (cached) return cached
+
+    const response = super.readPackageSync(location)
+    this.store[location] = response.result
+    return response
```

Comment:
if we only ever use the `result`, why return the entire response?

Reply:
@ljharb It's needed so that `readPackageUpSync` can report all locations which failed to yield a file - in the event that no fles were located.